# Python и интернет. Модуль requests

**План**:

1. Requests
4. Beautiful Soup
5. Краулеры

## Как устроен интернет

Оригинальные [конспект1](https://github.com/elmiram/2016learnpython/blob/master/7%20%D0%A1%D0%B5%D0%BC%D0%B8%D0%BD%D0%B0%D1%80%20-%20flask%20intro.ipynb) [конспект2](https://github.com/elmiram/2016learnpython/blob/master/8%20%D0%A1%D0%B5%D0%BC%D0%B8%D0%BD%D0%B0%D1%80%20-%20%D0%A1%D0%BD%D0%BE%D0%B2%D0%B0%20flask.ipynb)

__Веб-сервер__
* устройство vs. программа
* задача — получать запросы от других компьютеров или программ (клиентов, clients) и отправлять запрошенные данные
* основная функция — размещение сайтов (website hosting)

__Термины__
* *Запрос (request)* — сообщение от агента, желающего получить или разместить информацию
* *Ответ (response)* — ответное сообщение сервера с кодом состояния и информацией (HTML, изображения, загружаемые файлы и т. п.)
* *Протокол (protocol)* — набор правил, по которым составляются запросы и ответы
* *Сессия (session)* — установка соединения между агентом и сервером и последующая серия запросов и ответов
* *Гипертекст* — множество текстов, организованных в виде графа при помощи гиперссылок
* *Протокол HTTP (HyperText Transfer Protocol)* — протокол для передачи гипертекстовых данных (обычно в виде HTML)
* *URL (Uniform Resource Locator)*, веб-адрес ресурса — строка, представляющая собой уникальное имя, по которому можно найти в сети этот ресурс

__Адресация__

IP
* *IP-адрес* — (пока что) набор из 4 байт, присваиваемый каждому подключённому к сети устройству
* Некоторые IP-адреса уникальны, некоторые — нет (внутренние адреса в локальных сетях)
* Практически любой ресурс (например, сайт) можно получить по его IP-адресу (например, через браузер)
* Существуют зарезервированные адреса и диапазоны адресов, например, `127.0.0.1` — адрес данного устройства

Порт
* Каждый запрос обращается не просто к какому-то IP-адресу, а к некоторому порту на этом адресе
* Веб-сервер имеет 65 535 портов, пронумерованных начиная с 1
* Веб-сервер может прослушивать некоторые порты (listen to ports) и по-разному обрабатывать сообщения, поступившие на разные порты
* Если порт не прослушивается, сообщения на этот порт останутся без ответа


__URL__

`http__://__www.example.com__:__1234__/__directory1/file1.html`

`протокол__://__доменное имя или IP-адрес__:__порт__/__адрес файла на сервере`

* Порт указывать не обязательно: используются стандартные порты (HTTP — 80, FTP — 20 и т. п.)
* *DNS (Domain Name Servers)* — специальные сервера в сети, на которых хранятся таблицы соответствия между доменными именами и IP-адресами их серверов


__HTML и скрипты__
* страницы, содержимое (HTML-код) которых не изменяется, называются *статическими (static)*
* страницы, содержимое которых может быть разным в зависимости от введённых пользователем данных, времени, IP-адреса и т. п., называются *динамическими (dynamic)*
* динамические страницы создаются с помощью скриптов *на стороне сервера (server side scripts)*, написанных, например, на PHP или Python
* скрипт порождает HTML и посылает его пользователю
* пользователь не видит кода скрипта, выполняющегося на сервере
* *скрипт на стороне клиента (client side script)* — вставка в HTML на каком-то языке программирования (например, JavaScript), позволяющая странице вести себя интерактивно
* код клиентских скриптов посылается клиенту сервером вместе с HTML и не выполняется на сервере

__Питон и сайты__

Написать сайт на питоне значит написать такую программу, которая может работать веб-сервером или использоваться веб-сервером для порождения HTML-кода веб-страниц. Для этого существует несколько модулей, например, Django и Flask. Мы их рассмотрим на следующих парах.

А пока выясним, как данные из интернета доставать.

## Как выкачать интернет
Современный Интернет предоставляет большое количество языковых данных: электронные газеты и журналы, блоги, форумы, социальные сети и т.д. Например, можно найти в сети много-много текстов и собрать корпус, или найти все газетные статьи и блог-посты про какую-нибудь корпорацию и проанализировать тональность сообщений. Сейчас мы научимся заниматься выкачиванием страниц из интернета с помощью Python.

Для скачивания HTML-страниц в питоне есть несколько библиотек: **requests** и **urllib**.

## Requests

Можно почитать в [документации](https://requests.readthedocs.io/en/latest/) или [тут](https://realpython.com/python-requests/).

Допустим, мы хотим скачать главную страницу Хабра.

На самом деле, когда мы хотим открыть какую-то страницу в интернете, наш браузер отправляет на сервер **запрос** ("Привет, сервер! я хочу код страницы по вот такому адресу!"), а сервер затем отправляет ответ ("Привет! Вот код страницы: ...").
Чтобы получить страницу через питон, нужно сформировать **запрос** на сервер так же, как это делает браузер:

In [ ]:
import requests

In [ ]:
response = requests.get("https://habr.com/ru/")

В response теперь лежит ответ сервера. Это не просто HTML-код страницы, а еще дополнительная информация. Если мы просто выведем этот `response`, нам покажется только код (если статус $-$ 200, то все ок; вам, вероятно, хорошо известен статус 404).

In [ ]:
response

<Response [200]>

In [ ]:
response.status_code

200

А вот в атрибуте `text` уже лежит HTML-код

In [ ]:
print(response.text[:700])

<!DOCTYPE html>
<html lang="ru">

  <head>
    <title>Публикации &#x2F; Моя лента &#x2F; Хабр</title>
<link rel="image_src" href="/img/habr_ru.png" data-hid="2a79c45">
<link href="https://habr.com/ru/feed/" rel="canonical" data-hid="e3fa780">
<meta itemprop="image" content="/img/habr_ru.png">
<meta property="og:image" content="/img/habr_ru.png">
<meta property="og:image:width" content="1200">
<meta property="og:image:height" content="630">
<meta property="aiturec:image" content="/img/habr_ru.png">
<meta name="twitter:image" content="/img/habr_ru.png">
<meta property="vk:image" content="/img/habr_ru.png?format=vk">
<meta property="fb:app_id" content="444736788986613">
<meta property="fb:pages


Можно посмотреть, что еще можно достать из response.

Функция ```dir``` выдает список методов и параметров объекта.

In [ ]:
[i for i in dir(response) if not i.startswith("_")]

['apparent_encoding',
 'close',
 'connection',
 'content',
 'cookies',
 'elapsed',
 'encoding',
 'headers',
 'history',
 'is_permanent_redirect',
 'is_redirect',
 'iter_content',
 'iter_lines',
 'json',
 'links',
 'next',
 'ok',
 'raise_for_status',
 'raw',
 'reason',
 'request',
 'status_code',
 'text',
 'url']

- Кодировка:

In [ ]:
response.encoding

'utf-8'

- Заголовки (техническая информация):

In [ ]:
dict(response.headers)

{'Server': 'QRATOR',
 'Date': 'Thu, 06 Nov 2025 09:18:51 GMT',
 'Content-Type': 'text/html; charset=utf-8',
 'Transfer-Encoding': 'chunked',
 'Connection': 'keep-alive',
 'Keep-Alive': 'timeout=15',
 'Vary': 'Accept-Encoding, Accept-Encoding',
 'X-DNS-Prefetch-Control': 'off',
 'X-Frame-Options': 'SAMEORIGIN',
 'X-Download-Options': 'noopen',
 'X-Content-Type-Options': 'nosniff',
 'X-XSS-Protection': '1; mode=block',
 'ETag': 'W/"5fc08-DbEsZ8b3l3ccBYs59LnNsccwkqs"',
 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains',
 'X-Request-Id': 'cf98f5340a1b192fd0d99ddf1193b8a8',
 'X-Request-Geoip-Country-Code': 'US',
 'X-Request-Detected-Device': 'mobile',
 'Content-Encoding': 'gzip'}

- Адрес запроса:

In [ ]:
response.url

'https://habr.com/ru/feed/'

* Содержимое страницы:

In [ ]:
print(response.text[:300])

<!DOCTYPE html>
<html lang="ru">

  <head>
    <title>Публикации &#x2F; Моя лента &#x2F; Хабр</title>
<link rel="image_src" href="/img/habr_ru.png" data-hid="2a79c45">
<link href="https://habr.com/ru/feed/" rel="canonical" data-hid="e3fa780">
<meta itemprop="image" content="/img/habr_ru.png">
<meta 


Ура, всё на месте!

Но что всё это значит? Что такое HTML и как вообще из него доставать какую-то информацию?

Ответ: по **тегам**! Например, в куске HTML сверху есть теги `<title> </title>` (теги всегда обрамляют с двух сторон то, что находится под этим тегом). В `<title>` в данном случае лежит заголовок этой интернет-страницы.

Существует несколько вариантов, как достать что-то из определенного тега, например, достать заголовок:

1. регулярные выражения ([плохой вариант](https://stackoverflow.com/questions/590747/using-regular-expressions-to-parse-html-why-not))
2. специальные библиотеки питона, например, BeautifulSoup или [lxml](https://lxml.de/) (хороший вариант)

## BeautifulSoup

[Документация](https://www.crummy.com/software/BeautifulSoup/bs4/doc/)

Код страницы парсится как иерархия тегов (как они есть в html коде, один вложен в другой) и можно искать элементы с помощью разных методов, использовать атрибуты и т.д.

In [ ]:
from bs4 import BeautifulSoup

Сначала инициализируем объект `BeautifulSoup`. Потом применим метод `find` и в скобочках укажем теги, по которым ищем. У некоторых тегов в HTML (как и в нашем случае) бывает еще и `class` и какие-нибудь еще атрибуты. Такие вещи мы задаем словариком.

Этот запрос вернёт нам только первый заголовок. То есть первое вхождение такого тега в нашем HTML файле.

In [ ]:
soup = BeautifulSoup(response.text, 'html.parser')  # инициализируем (создаем) soup

post = soup.find('h2', {'class': 'tm-title tm-title_h2'})
print(post.prettify())
print(post.get_text(), end="\n\n\n")

<h2 class="tm-title tm-title_h2" data-test-id="articleTitle">
 <!--[-->
 <a class="tm-title__link" data-article-link="true" data-test-id="article-snippet-title-link" href="/ru/companies/otus/articles/958458/">
  <span>
   OWASP TOP Ten: 10 болей ИБ для машинного обучения
  </span>
 </a>
 <!--]-->
</h2>

OWASP TOP Ten: 10 болей ИБ для машинного обучения




Но мы хотим получить все заголовки постов! Метод `find_all` возвращает массив всех элементов с тегом указанным в скобках. По нему можно итерироваться.

In [ ]:
for post in soup.find_all('h2', {'class': 'tm-title tm-title_h2'})[:3]:
    print(post.get_text())
    print(post.prettify())

    print('-- ' * 10)  # для красоты

Как мы автоматизировали код-ревью благодаря связке Aider + LLM
<h2 class="tm-title tm-title_h2" data-test-id="articleTitle">
 <!--[-->
 <a class="tm-title__link" data-article-link="true" data-test-id="article-snippet-title-link" href="/ru/companies/fix_price/articles/963658/">
  <span>
   Как мы автоматизировали код-ревью благодаря связке Aider + LLM
  </span>
 </a>
 <!--]-->
</h2>

-- -- -- -- -- -- -- -- -- -- 
Часть ПК с Windows загружаются в режиме восстановления BitLocker после последних апдейтов
<h2 class="tm-title tm-title_h2" data-test-id="articleTitle">
 <!--[-->
 <a class="tm-title__link" data-article-link="true" data-test-id="article-snippet-title-link" href="/ru/news/963634/">
  <span>
   Часть ПК с Windows загружаются в режиме восстановления BitLocker после последних апдейтов
  </span>
 </a>
 <!--]-->
</h2>

-- -- -- -- -- -- -- -- -- -- 
Java Digest #30
<h2 class="tm-title tm-title_h2" data-test-id="articleTitle">
 <!--[-->
 <a class="tm-title__link" data-article-link="tr

## Задание на семинар 1

А что если мы хотим зайти еще глубже по дереву тегов и, например, для каждого заголовка поста найти никнейм юзера, который написал этот пост, и время написания поста?

Для этого надо снова зайти в просмотор кода страницы и увидеть, что там проиcходит что-то такое:

(Заодно обратите внимание, как пишутся комменты в HTML)

```
<article class="tm-articles-list__item" data-navigatable="" id="771476" tabindex="0">
 <div class="tm-article-snippet tm-article-snippet">
  <div class="tm-article-snippet__meta-container">
   <div class="tm-article-snippet__meta">
    <span class="tm-user-info tm-article-snippet__author">
     <a class="tm-user-info__userpic" href="/ru/users/darinka666/" title="darinka666">
      <div class="tm-entity-image">
       <img alt="" class="tm-entity-image__pic" height="24" src="https://assets.habr.com/habr-web/img/avatars/093.png" width="24"/>
      </div>
     </a>
     <span class="tm-user-info__user tm-user-info__user_appearance-default">
      <a class="tm-user-info__username" href="/ru/users/darinka666/">
       darinka666
       <!-- -->
      </a>
      <span class="tm-article-datetime-published">
       <time datetime="2023-11-02T09:22:25.000Z" title="2023-11-02, 12:22">
        11 минут назад
       </time>
      </span>
     </span>
    </span>
   </div>
   <!-- -->
  </div>
  <h2 class="tm-title tm-title_h2">
   <a class="tm-title__link" data-article-link="true" data-test-id="article-snippet-title-link" href="/ru/companies/mts_ai/articles/771476/">
    <span>
     Обзор Llemma: новая математическая open-source модель
    </span>
   </a>
  </h2>
  <div class="tm-article-snippet__stats">
   <div class="tm-article-complexity tm-article-complexity_complexity-medium">
    <span class="tm-svg-icon__wrapper tm-article-complexity__icon">
     <svg class="tm-svg-img tm-svg-icon" height="24" width="24">
      <title>
       Уровень сложности
      </title>
      <use xlink:href="/img/megazord-v28.2fb1b1c1..svg#complexity-medium">
      </use>
     </svg>
    </span>
    <span class="tm-article-complexity__label">
     Средний
    </span>
   </div>
```
(и так далее; часть вывода обрезана: например, нет закрывающего тега `</article>`)

In [ ]:
username = soup.find('a', {'class': 'tm-user-info__username'})
username.get_text()

time = soup.find('time')
time.get_text()

'10 минут назад'

In [ ]:
for article in soup.find_all('article'):
  username = article.find('a', {'class': 'tm-user-info__username'}).get_text()
  time = article.find('time').get_text()
  title = article.find('h2', {'class': 'tm-title tm-title_h2'}).get_text()
  print(title, username, time, sep='\n', end='\n\n')

OWASP TOP Ten: 10 болей ИБ для машинного обучения
Andrey_Biryukov 
10 минут назад

Взгляд безопасника на ежегодный отчет Github Octoverse 2025
vasily2015 
12 минут назад

Поднимаем внутренний мессенджер и VoIP-сервер без нарушения законодательства
ITSumma 
13 минут назад

SysAdmin 2025: российские сисадмины делятся опытом
RingoMDM 
22 минуты назад

Ansible Inventory: полное руководство от localhost до AWS
net0pyr 
26 минут назад

Как некачественные данные подтачивают способности нейросетей и что с этим делать
BiktorSergeev 
27 минут назад

От проблемы к запуску: как превратить результаты исследований в реальные продукты
ReasonVoice 
29 минут назад

Perplexity вложит $400 млн в Snap, чтобы внедрить свой ИИ-поиск в Snapchat
Travis_Macrif 
30 минут назад

Купил увлекательный квест за 0,00896 BTC
Ares_ekb 
32 минуты назад

Вышла Kimi K2-Thinking. Новый ИИ из Китая доступен бесплатно
runaway_llm 
33 минуты назад

Проектирование и информационное моделирование мостовых сооружений в nanoCAD. О

## Задание на семинар 2

Скачать главную страницу Яндекс.Погоды и
    
- распечатать сегодняшнюю температуру и облачность
- распечатать время восхода и заката
- погоду на завтра

Простая версия: вместо Яндекс.Погоды возьмите [этот](https://simple-weather-website.netlify.app/) сайт и получите температуру, влажность и скорость ветра.

## Краулеры

## Что такое краулеры?

Краулеры $-$ это боты / программы, которые "ползают" (*crawl*) по страницам сайта и собирают информацию. Все чаще использование таких программ запрещается правилами пользования сайтами, поэтому это формально нехорошо. Но так продолжают делать и это надо уметь. Запрещают по 2 основным причинам: не хотят делиться данными и боятся, что вы уроните сервер (если сайт $-$ маленький, а сервер $-$ не очень, то это довольно легко).

Поэтому нужно собирать данные аккуратно, чтобы:
1. вас не заблокировали по IP,
2. вы не навредили серверу.

Вместо отдельных запросов для этого лучше создать сессию, которая позволит хранить информацию между запросами и поддерживать то же соединение и не создавать каждый раз все заново, что влияет на производительность.



In [ ]:
from pprint import pprint # библиотека для красивого вывода

In [ ]:
session = requests.session()

Попробуем сделать запрос, просто вместо requests.get мы пишет session.get.

In [ ]:
response = session.get('https://eksmo.ru/khudozhestvennaya-literatura/fantastika/')

In [ ]:
response

<Response [403]>

Посмотреть на `headers` запроса

In [ ]:
pprint(dict(response.headers))

{'Cache-Control': 'no-cache',
 'Connection': 'keep-alive',
 'Content-Encoding': 'gzip',
 'Content-Type': 'text/html',
 'Date': 'Thu, 06 Nov 2025 14:16:26 GMT',
 'ETag': 'W/"690c9f9e-4a05e"',
 'Expires': 'Wed, 06 Nov 2024 14:16:26 GMT',
 'Keep-Alive': 'timeout=15',
 'Last-Modified': 'Thu, 06 Nov 2025 13:16:14 GMT',
 'Server': 'QRATOR',
 'Strict-Transport-Security': 'max-age=31536000;',
 'Transfer-Encoding': 'chunked',
 'Vary': 'Accept-Encoding',
 'X-Bitrix-Composite': 'Nginx (file)',
 'X-Content-Type-Options': 'nosniff',
 'X-Frame-Options': 'SAMEORIGIN',
 'X-XSS-Protection': '1; mode=block',
 'n_cc': '0',
 'n_score': '0'}


### Стратегии сбора данных


По сути краулеры выполняют сбор страниц (их HTML) как мы это делали на прошлом занятии, но делают они это циклами (или циклами циклов). Можно выделить разные стратегии сбора данных:
    
**По типу навигации**

1. Все страницы со ссылками имеют удобные номера ("https://ficbook.net/fanfiction/no_fandom/originals?p=2"), обычно просто `p=<число>` или `page=<число>`. В этом случае вам нужно просто подставлять цифры подробнее про параметры передаваемые в ссылке можно посмотреть [здесь](https://en.wikipedia.org/wiki/Query_string).
2. Страницы называются как-то не структурированно (например, по названиям блоков). Тут нужно собирать ссылки на эти страницы и потом по ним ходить и собирать конечные странички.
3. Все расположено на одной страничке и догружается с использованием [WebSocket](https://en.wikipedia.org/wiki/WebSocket) или других технологий, при адрес в адресной строке никак не изменяется, данные могут догружаться на сайт автоматически по мере скролла страницы.

**По скорости обновления**

1. Если сайт довольно статичный по контенту (медленно появляются и удаляются материалы), то можно сначал собрать ссылки, а потом по ним ходить.
2. Если сайт очень динамичный по контенту (например, объявления на крупном сайте), вам нужно при получении страничкии ссылок сразу их обходить, а потом переходить к следующей, потому что ко времени получения исчерпывающего списка ссылок по сайту многие будут уже удалены или недоступны.



## Блокировки и способы их обхода

Для того, чтобы предотвратить автоматический сбор информации с некого сайта, применяются различные инструменты, которые определяют роботов и блокируют запросы с адресов, которые были классифицированы как роботы. Чтобы не заблокировали домашний/учебный IP, лучше сразу задуматься об этих мерах и предотвратить возможные проблемы. Кстати, Википедия не блокирует и можно спокойно скачивать без каких-либо проблем.

Чтобы их обойти, можно попробовать несколько инструментов:

1. изобразить браузер - при запросе отправляется информация о том, из какого приложения пришел запрос (например, Google Chrome), запросы сделанные из браузера больше похожи на человеческие, для этого нужно задать `user-agent` в параметрах (а его выбирать случайно с помощью `fake_useragent`).
1. `time.sleep(x)` - задержка между запросами, чтобы слишком большая скорость запросов не показалась подозрительной или ваши запросы не уронили сервер небольшого ресурса (например, региональной газеты).
2. `time.sleep(<случайный промежуток времени>)` - это более хитрая версия, когда время задержки - это случайное число из некоторого отрезка (модуль `random`).
4. использовать прокси - существуют ресурсы с бесплатными списками открытых прокси, через которые можно пропускать ваш запрос и сервер будет думать, что запросы приходят из разных мест (anonymous и elite классы прокси) или использовать анонимизированные сети к примеру сеть [Tor](https://en.wikipedia.org/wiki/Tor_(network)) и аналоги.

При работе в Google Colab зачастую помогает перезапускать среду: код в Colab выполняется на внешних серверах, соответственно, при перезапуске будет отправлять запросы уже с другого сервера. Хотя иногда это приводит и к обратным проблемам: не все сайты доступны вне России, например (как и не все доступны из России))

### Притвориться нормальным браузером

In [ ]:
!pip install fake_useragent

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 161.7/161.7 kB 3.9 MB/s eta 0:00:00


In [ ]:
from fake_useragent import UserAgent

Можно настроить так, чтобы не проверять безопасность соединения, что иногда вызывает ошибки. Но это можно делать с сайтами, которым вы доверяете.

In [ ]:
ua = UserAgent()

In [ ]:
headers = {'User-Agent': ua.random}
# другой способ
# headers = {'User-Agent': ua.chrome}
print(headers)
response = session.get('https://eksmo.ru/khudozhestvennaya-literatura/fantastika/', headers=headers)

{'User-Agent': 'Mozilla/5.0 (iPhone; CPU iPhone OS 18_4_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) CriOS/135.0.7049.53 Mobile/15E148 Safari/604.1'}


In [ ]:
response

<Response [403]>

### Пауза между запросами

In [ ]:
import time
from datetime import datetime

In [ ]:
for _ in range(5):
    response = session.get('https://eksmo.ru/khudozhestvennaya-literatura/fantastika/')
    print(datetime.now(), response)
    time.sleep(1)

2025-11-06 14:23:28.160152 <Response [200]>
2025-11-06 14:23:29.303595 <Response [403]>
2025-11-06 14:23:30.445944 <Response [403]>
2025-11-06 14:23:31.591222 <Response [403]>
2025-11-06 14:23:32.735247 <Response [403]>


Некоторые сайты замечают подозрительно регулярные запросы. `random.uniform` позволяет получить случайное число из отрезка и сделать наши запросы менее подозрительными.

In [ ]:
import random

In [ ]:
random.uniform(1, 3)

2.2396444847550256

In [ ]:
response = session.get('https://eksmo.ru/khudozhestvennaya-literatura/fantastika/')

In [ ]:
for _ in range(5):
    response = session.get('https://eksmo.ru/khudozhestvennaya-literatura/fantastika/')
    print(datetime.now(), response)
    time.sleep(random.uniform(1.1, 5.2))

2025-11-06 14:22:30.180634 <Response [200]>
2025-11-06 14:22:34.902091 <Response [200]>
2025-11-06 14:22:39.179946 <Response [200]>
2025-11-06 14:22:41.484531 <Response [200]>
2025-11-06 14:22:46.625367 <Response [200]>


А всё, поздно уже.. (Статус 403 - Forbidden: сервер получил наш запрос, но отказался на него отвечать)

В идеале все перечисленные способы нужно использовать сразу, особенно если вы работаете локально, иначе сайт мжет вас запомнить и не пускать, например, несколько часов.

### Подключение через прокси

Прокси-сервер — это дополнительное звено между вами и интернетом, через него пойдет подключение и сайт не будет знать, что это вы посылаете запрос.

1. Адреса прокси можно взять со специальных сайтов: например, [https://hideip.me/ru/proxy/httplist](https://hideip.me/ru/proxy/httplist).
2. Потом проверить, что они рабочие, прежде чем использовать: [https://checkerproxy.net/](https://checkerproxy.net/).

In [ ]:
proxy = {"https://": "https://176.213.141.107:8080"}

response = session.get('https://eksmo.ru/khudozhestvennaya-literatura/fantastika/', proxies=proxy)

In [ ]:
response

<Response [403]>

## Пример

Давайте обкачаем немного новостей с сайта вышки.

1. Страницы имеют вид "https://www.hse.ru/news/page1.html", поэтому можно просто идти циклом.
2. Достанем дату публикации, заголовок, краткое описание (из станицы со списком новостей), текст полной статьи и метки (из самой страницы новости)
3. Сохраним в датафрейм.

In [ ]:
from bs4 import BeautifulSoup
import re

### **Шаг 1. Найти страницы**

Посмотрим, как устроены новости и скачаем одну страницу:

In [ ]:
page_number = 3
url = f'https://www.hse.ru/news/page{page_number}.html'
req = session.get(url, headers={'User-Agent': ua.random})
page = req.text

Распарсим с помощью `BeautifulSoup`:

In [ ]:
soup = BeautifulSoup(page, 'html.parser')

Найдем отдельные посты:

In [ ]:
news = soup.find_all('div', {'class': 'post'})

In [ ]:
len(news)

10

In [ ]:
print(news[0].prettify())

<div class="post post_first">
 <div class="post__extra">
  <div class="post-meta">
   <div class="post-meta__filter-ico rubric-white rubric_32">
    <svg height="70" viewbox="0 0 70 70" width="70" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
     <rect clip-path="url(#hse-mask)" fill="#0941AF" height="70" width="70" x="0" y="0">
     </rect>
     <image class="svg" clip-path="url(#hse-mask)" height="70" src="/f/src/global/i/mask-rubric-cut.png" width="70" xlink:href="/f/src/global/i/mask-rubric-cut.png"/>
    </svg>
    <span class="filter-ico filter-ico_mask">
     <img height="70" src="/f/src/global/i/mask-pic.png" width="70"/>
    </span>
   </div>
   <div class="post-meta__date">
    <div class="post-meta__day">
     30
    </div>
    <div class="post-meta__month">
     окт
    </div>
    <div class="post-meta__year">
     2025
    </div>
    <div class="post-meta__visitors">
    </div>
   </div>
  </div>
 </div>
 <div class="post__content">
  <div

Найдем заголовок-ссылку и запомним текст заголовка:

In [ ]:
title_obj = news[0].find('a')
title_obj

<a class="link link_dark2 no-visited" href="/news/science/1097453616.html">«Развернуть обсуждение политики в области высшего образования в доказательное русло»</a>

In [ ]:
title = title_obj.text
title

'«Развернуть обсуждение политики в области высшего образования в доказательное русло»'

Достанем свойства этой ссылки (куда ведет, `class`):

In [ ]:
attrs = title_obj.attrs
attrs

{'href': '/news/science/1097453616.html',
 'class': ['link', 'link_dark2', 'no-visited']}

Достанем саму ссылку:

In [ ]:
href = title_obj.attrs['href']
href

'/news/science/1097453616.html'

Достанем текст новости:

In [ ]:
short_text = news[0].find('div', {'class': 'post__text'}).text
short_text

'29 октября в НИУ ВШЭ открылась XVI Международная конференция исследователей высшего образования (ИВО) на тему «Высшее образование: между частным и общественным благом». Для участия в конференции зарегистрировались более 600 человек из 32 регионов России и семи зарубежных стран, поступило рекордное число заявок на выступления с докладами — 242, из которых было принято 88.'

Достанем день, месяц, год публикации:

In [ ]:
pub_day = news[0].find('div', {'class': 'post-meta__day'}).text
pub_day

'29'

In [ ]:
pub_month = news[0].find('div', {'class': 'post-meta__month'}).text
pub_month

'окт'

In [ ]:
pub_year = news[0].find('div', {'class': 'post-meta__year'}).text
pub_year

'2025'

### **Шаг 2. Научиться парсить страничку самой новости**

Скачаем и распарсим полученную ссылку:

In [ ]:
url = 'https://www.hse.ru/news/science/1099260087.html'

req = session.get(url, headers={'User-Agent': ua.random})
page = req.text

soup = BeautifulSoup(page, 'html.parser')

Сохраним текст, распечатаем кусочек:

In [ ]:
full_text = soup.find('div', {'class': 'post__content'}).text
full_text[:200]

'«Развернуть обсуждение политики в области высшего образования в доказательное русло»29 октября в НИУ ВШЭ открылась XVI Международная конференция исследователей высшего образования (ИВО) на тему «Высше'

Найдем теги, которые присвоены статье:

In [ ]:
meta = soup.find('div', {'class': 'articleMeta'})

tags = meta.find_all('a', {'class': 'tag'})
tags = [t.text for t in tags]
tags

['исследования и аналитика', 'репортаж о событии']

### **Шаг 3. Оформляем нормально в функции**

Сделаем словарь соответствий имени месяца и его номера:

In [ ]:
months = {
    value: key + 1
    for key, value in enumerate(
        ['янв', 'фев', 'мар', 'апр', 'мая', 'июн', 'июл', 'авг', 'сен', 'окт', 'ноя', 'дек']
    )
}

Парсим информацию из страницы со списком новостей (блок одной новости):

In [ ]:
def parse_news_page_block(one_block):
    block = {}
    a = one_block.find('a')
    block['title'] = a.text
    block['href'] = a.attrs['href']
    block['short_text'] = one_block.find('div', {'class': 'post__text'}).text
    block['pub_day'] = int(one_block.find('div', {'class': 'post-meta__day'}).text)
    block['pub_month'] = months[one_block.find('div', {'class': 'post-meta__month'}).text]
    block['pub_year'] = int(one_block.find('div', {'class': 'post-meta__year'}).text)
    return block

Парсим отдельную страницу новости:

In [ ]:
def parse_one_article(block):
    url_one = 'http://www.hse.ru' + block['href']
    req = session.get(url_one, headers={'User-Agent': ua.random})
    page = req.text
    soup = BeautifulSoup(page, 'html.parser')
    block['full_text'] = soup.find('div', {'class': 'post__content'}).text
    meta = soup.find('div', {'class': 'articleMeta'})
    tags = meta.find_all('a', {'class': 'tag'})
    block['tags'] = [t.text for t in tags]
    return block

Регулярное выражение для того, чтобы достать ID новости и не повторяться:

In [ ]:
regex_hse_id = re.compile('/([0-9]*?).html')

Обработать N-ую страницу новостей:

In [ ]:
def get_nth_page(page_number):
    # скачиваем
    url = f'https://www.hse.ru/news/page{page_number}.html'
    req = session.get(url, headers={'User-Agent': ua.random})
    page = req.text
    soup = BeautifulSoup(page, 'html.parser')

    # находим новости
    news = soup.find_all('div', {'class': 'post'})

    # идем по новостям и обрабатываем их
    blocks = []
    for n in news:
        try:
            blocks.append(parse_news_page_block(n))
        except Exception as e:
            print(e)

    # идем по отдельным статьям и достаем информацию
    result = []
    for b in blocks:
        if b['href'].startswith('/'):
            idx = regex_hse_id.findall(b['href'])[0]
            try:
                res = parse_one_article(b)
                res['hse_id'] = idx
                result.append(res)
            except Exception as e:
                print(e)

    # возвращаем найденную информацию
    return result

### **Шаг 4. Сохраняем в датафрейм**


In [ ]:
from tqdm.auto import tqdm
import pandas as pd

Напишем функцию, куда передаем количество страниц и она выполняет все нужные действия:

In [ ]:
def run_all(n_pages):
    blocks = []
    for i in tqdm(range(n_pages)):
        blocks.extend(get_nth_page(i+1))

    return blocks

Запускаем на 20 первых страниц:

In [ ]:
blocks = run_all(3)

df = pd.DataFrame(blocks)

  0%|          | 0/3 [00:00<?, ?it/s]

In [ ]:
df.sample(3)

title  \
2   Когда вирус наступает на мину: найден древний ...   
19  Исчезнувший сигнал: как солнечная активность з...   
5   Магистрант Вышки — семикратный чемпион России ...   

                             href  \
2   /news/science/1099260087.html   
19  /news/science/1097022005.html   
5      /news/life/1098197960.html   

                                           short_text  pub_day  pub_month  \
2   Когда вирус попадает в клетку, он вмешивается ...        5         11   
19  Исследователи из НИУ ВШЭ и ИКИ РАН проанализир...       28         10   
5   В октябре Михаил Симонов вновь подтвердил титу...        1         11   

    pub_year                                          full_text  \
2       2025  Когда вирус наступает на мину: найден древний ...   
19      2025  Исчезнувший сигнал: как солнечная активность з...   
5       2025  Магистрант Вышки — семикратный чемпион России ...   

                                                 tags      hse_id  
2   [публикации, исследования и аналитика, програм...  1099260087  
19             [публикации, исследования и аналитика]  1097022005  
5                       [достижения, студенты, спорт]  1098197960

Посмотрим на 10 самых популярных тегов

In [ ]:
from collections import Counter

In [ ]:
tags = [tag for tags in df["tags"] for tag in tags]

In [ ]:
for title, counts in sorted(Counter(tags).items(), key=lambda x: -x[-1])[:10]:
    print(counts,"\t", title)

9 	 исследования и аналитика
7 	 репортаж о событии
5 	 публикации
4 	 международное сотрудничество
3 	 центры превосходства
2 	 новое в ВШЭ
2 	 дополнительное образование
2 	 достижения
2 	 студенты
2 	 дискуссии


Посмотрим, сколько публикаций по месяцам

In [ ]:
df.groupby("pub_month")["hse_id"].count()

pub_month
10    16
11     7
Name: hse_id, dtype: int64

## Задание на семинар

Сделайте краулер фильмов аналогичный тому, что выше.

У вас есть список на [Letterboxd](https://letterboxd.com/) c перечислением всех фильмов из книжки *1001 Movies You Must See Before You Die*: https://letterboxd.com/peterstanley/list/1001-movies-you-must-see-before-you-die/detail/.

1. Все страницы выглядят как `https://letterboxd.com/peterstanley/list/1001-movies-you-must-see-before-you-die/detail/page/<номер>/`, поэтому снова можно идти циклом.
2. Достанем номер фильма, его название, год выпуска со страницы списка.
3. Достанем режиссера, слоган (если есть) и краткое описание с самой страницы фильма.
4. Сохраним в датафрейм.

## *Решение задания на семинар 1

In [ ]:
for post in soup.find_all("article"):
    try:  # чтобы исключить случаи, когда нет юзернейма / времени / заголовка
        user_name = post.find('a', {'class': 'tm-user-info__username'}).get_text()
        time_article = post.find('time').get_text()
        header_article = post.find('h2', {'class': 'tm-title tm-title_h2'}).get_text()
        print('Name:', user_name)
        print('Time:', time_article)
        print('Header:', header_article, end='\n\n')
    except AttributeError:
        # ошибка AttributeError возникает, когда от объекта типа None вызывают метод .get_text()
        pass

Name: ksrepin 
Time: 1 минуту назад
Header: Как мы автоматизировали код-ревью благодаря связке Aider + LLM

Name: maybe_elf 
Time: 1 минуту назад
Header: Часть ПК с Windows загружаются в режиме восстановления BitLocker после последних апдейтов

Name: roma00712 
Time: 6 минут назад
Header: Java Digest #30

Name: alyonayurchenko 
Time: 15 минут назад
Header: РосНОУ вошёл в Университетский консорциум исследователей больших данных

Name: Realife 
Time: 17 минут назад
Header: Реставрация, которая меня сломала: Почему убрать смех из Скуби-Ду сложнее, чем сделать ремастер Тома и Джерри в 2к

Name: tvaleev 
Time: 18 минут назад
Header: Открыт доступ к ОС Аврора Developer Preview 5.2.0 для раннего тестирования

Name: LKamrad 
Time: 21 минуту назад
Header: Перед вами первый «торговый автомат» по продаже крепкого алкоголя. Вы не поверите, но ему уже почти три века

Name: maybe_elf 
Time: 21 минуту назад
Header: Canon установила 32-мегапиксельный сенсор и добавила поддержку видео 7K в камеру EOS R